Board listing fixes

57300 opened the pull request:
Re-applies #69904 and re-fixes #70614

----

* Error out if multiple boards use the same name
* Ignore duplicate roots in list_boards/hardware

---

(edit by aescolar) Fixes: #69329

tejlmand committed:
scripts: remove board name from the qualifiers in list_boards.py

Fixes: #69329

The board name was printed as part of printing board qualifiers because
those was being concatenated in the `board_v2_qualifiers()` method.

Keep the qualifiers separated from the board name and let the caller
concatenate the strings when required.

Completion scripts are also updated to handle the corrected behaviour.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>
(cherry picked from commit 66b475a3aa4d73f4ddd5be5a7a1e0f3c7028e539)

fabiobaltieri commented:
I would restore the original comment for the revert of the revert, maybe just mention that it's a second attempt in the message.

57300 committed:
scripts: Ignore duplicate roots in list_boards/hardware

When iterating over `--arch-root`, `--board-root`, and `--soc-root`,
treat them as collections of absolute paths with no repeats, to ensure
that no input root has to be handled more than once.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
cmake: boards: Error out if multiple boards use the same name

Detecting this requires `boards.cmake` being able to handle multi-line
output from `list_boards.py`. Implement a similar line reading loop to
the one used in `hwm_v2.cmake`.

Failing to handle this could result in an incorrectly parsed list of
valid board qualifiers. Here's the expected list for `nrf52_bsim`:

  - "native"

and here's what would happen if two copies of that board were found:

  - "native\nNAME"
  - "nrf52_bsim"
  - "native"

Instead of that, there would now be a proper error message listing all
board directories which contain `nrf52_bsim`.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 commented:
> I would restore the original comment for the revert of the revert, maybe just mention that it's a second attempt in the message.

I think that's reasonable. Replaced with a cherry-pick.

57300 commented:
The bug observed in #70614 has two facets, both of which should now be fixed.

tejlmand commented on cmake/modules/boards.cmake:
```diff
@@ -195,6 +195,24 @@ endif()
+      endif()
+      string(FIND "${ret_board}" "\n" idx REVERSE)
+    endwhile()
+    message(FATAL_ERROR "Multiple boards named '${BOARD}' found in:${board_dirs}")
```
I would prefer we detect this at the `list_boards.py` level, especially considering that in future we need to also support this: https://github.com/zephyrproject-rtos/zephyr/issues/69548

as well as ensuring that other commands, like `west boards` detect such errors as well.

However, as immediate fix, then this proposal is accepted and can be cleaned up when implementing https://github.com/zephyrproject-rtos/zephyr/issues/69548.

57300 replied:
> I would prefer we detect this at the `list_boards.py` level

Agreed, and I'd like to get around to this as soon as I can.

I've already started work on refactoring the Python scripts, which would make this easier to detect at Python level, and would include performance improvements among other things.
